pydantic/_internal/_validate_call.py

```diff
@@ -97,3 +102,77 @@ def __call__(self, *args: Any, **kwargs: Any) -> Any:
         if self.__return_pydantic_validator__:
             return self.__return_pydantic_validator__(res)
         return res
+
+
+class ValidateCallInfo(TypedDict):
+    validate_return: bool
+    config: ConfigDict | None
+    function: Callable[..., Any]
+    local_namspace: dict[str, Any] | None
+
+
+def _is_wrapped_by_validate_call(obj: object) -> bool:
+    return hasattr(obj, '__pydantic_validate_call_info__')
+
+
+def collect_validate_call_info(namespace: dict[str, Any]):
+    validate_call_infos = {}
+    for name, func in namespace.items():
+        if _is_wrapped_by_validate_call(func):
+            validate_call_infos[name] = func.__pydantic_validate_call_info__
+
+    return validate_call_infos
+
+
+def _replicate_validate_call(info: ValidateCallInfo) -> Callable[..., Any]:
+    """When normally calling `validate_call`, we use the namespace of the frame that called it as local_ns.
+    This function mock that behavior by calling `validate_call` inside a new frame where we have copied all
+    local variables into.
+    """
+    namespace = info['local_namspace']
+
+    from ..validate_call_decorator import _validate_call_with_namespace
+
+    return _validate_call_with_namespace(
+        func=info['function'],
+        config=info['config'],
+        validate_return=info['validate_return'],
+        local_ns=namespace,
+    )
+
+
+def _copy_func(function: Callable[..., Any]):
+    @wraps(function)
+    def wrapper(*args, **kwargs):
+        return function(*args, **kwargs)
+
+    return wrapper
+
+
+def update_generic_validate_call_info(model: type[BaseModel]):
+    """Called when a generic model is subscripted."""
+    origin = model.__pydantic_generic_metadata__['origin']
+    typevars_map: dict[Any, Any] | None = _generics.get_model_typevars_map(model)
+
+    if not origin:
+        return
+
+    for func_name, info in origin.__pydantic_validate_call_infos__.items():
+        info = info.copy()
+        function = info['function'] = _copy_func(info['function'])
```

Comment:
I'm wondering if the "copy"/wrapping of the function is necessary. Afaiu, it is only necessary because we reassigned `__annotations__` to `function`. I don't know if it's possible, without too many changes, to avoid doing so, but probably not.

Reply:
I have a different approach in the new commit. Now the underlying `raw_function` will be the same object, but `__annotations__` and `__qualname__` of wrapper will be properyly updated, which is tested by `test_generic_wraps`.